### PR TITLE
[Core] Fix tracking of model forward time to the span traces in case of PP>1

### PR DESCRIPTION
**Pull request by sfc-gh-mkeralapura**
This is a quick follow up to https://github.com/vllm-project/vllm/pull/7089. In that PR we left the PP>1 unsupported for the model forward time. Fixing that here.

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by comaniac** [comment]
cc @rkooo567

File: vllm/worker/model_runner.py [hunk 1]
```diff
@@ -1551,6 +1551,21 @@ def execute_model(
 
         # Compute the logits in the last pipeline stage.
         if not get_pp_group().is_last_rank:
+            if (self.is_driver_worker
+                    and hidden_or_intermediate_states is not None
+                    and isinstance(hidden_or_intermediate_states,
+                                   IntermediateTensors)
+                    and self.observability_config is not None
+                    and self.observability_config.collect_model_forward_time):
+                model_forward_end.synchronize()
```
Comment (comaniac) [inline vllm/worker/model_runner.py:1560]: Just to confirm, does this sync hurt the performance in normal runs? I guess not because you won't have observability config?

File: vllm/worker/model_runner.py [hunk 2]
```diff
@@ -1732,7 +1752,7 @@ def forward(
                                                       **kwargs)
         if intermediate_tensors is not None:
             for key in intermediate_tensors.tensors:
-                if key != "model_execute_time":
+                if key != "model_execute_time" and key != "model_forward_time":
```
Comment (comaniac) [inline vllm/worker/model_runner.py:1755]:
```suggestion
                if key not in ("model_execute_time", "model_forward_time"):
```

File: vllm/worker/model_runner.py (hunk 1 above)
Comment (sfc-gh-mkeralapura) [reply vllm/worker/model_runner.py:1560]: Yes, this path won't be taken because observability_config.collect_model_forward_time won't be set.

**Review by comaniac — APPROVED** [review]
LGTM. Leave to @rkooo567

File: vllm/worker/model_runner.py [hunk 3]
```diff
@@ -1551,6 +1551,21 @@ def execute_model(
...
+                                   IntermediateTensors)
+                    and self.observability_config is not None
+                    and self.observability_config.collect_model_forward_time):
+                model_forward_end.synchronize()
+                model_forward_time = model_forward_start.elapsed_time(
+                    model_forward_end)
+                orig_model_forward_time = 0.0
+                if intermediate_tensors is not None:
+                    orig_model_forward_time = intermediate_tensors.tensors.get(
+                        "model_forward_time", torch.tensor(0.0)).item()
```
Comment (rkooo567) [inline vllm/worker/model_runner.py:1566]: why do we store this to tensor? any way to just use cpu data?

**Review by rkooo567 — COMMENTED** [review]
QQ: is it possible to add unit tests for this?

File: vllm/worker/model_runner.py (hunk 3 above)
Comment (sfc-gh-mkeralapura) [reply vllm/worker/model_runner.py:1566]: as far as I can tell, the only thing passed from the pipeline workers is the IntermediateTensors in serialized form. Hence added it to that. Is there a wrapper object of some form that holds these ?

**Comment by sfc-gh-mkeralapura** [comment]
> QQ: is it possible to add unit tests for this?

Let me look into that. There is one just the reporting of these metrics and two the PP>1 case. Let me look into see how doable these are. I will circle back later in the day.

**Comment by sfc-gh-mkeralapura** [comment]
> QQ: is it possible to add unit tests for this?

I could not figure out how to get a unittest for this part of the worker. I instead added a test in the overall tracing test to test for these detailed trace data. It does not test for the pp>1 case though.

please take a look.

File: tests/tracing/test_tracing.py [hunk 4]
```diff
@@ -114,5 +114,71 @@ def test_traces(trace_service):
...
+    assert metrics.scheduler_time > 0
     assert attributes.get(
         SpanAttributes.LLM_LATENCY_TIME_IN_SCHEDULER) == metrics.scheduler_time
+    # Model forward and model execute should be none, since detailed traces is
+    # not enabled.
+    assert metrics.model_forward_time is None
+    assert metrics.model_execute_time is None
+
+
+def test_traces_with_detailed_steps(trace_service):
```
Comment (rkooo567) [inline tests/tracing/test_tracing.py:126]: Add the same test with pp= 2?

File: vllm/worker/model_runner.py (hunk 3 above)
Comment (sfc-gh-mkeralapura) [reply vllm/worker/model_runner.py]: Done. It looks like the worker serializes a Dict[Str, Any], so it can serialize floats too.

File: tests/tracing/test_tracing.py (hunk 4 above)
Comment (sfc-gh-mkeralapura) [reply tests/tracing/test_tracing.py]: It was a bit more involved, but done.

**Comment by sfc-gh-mkeralapura** [comment]
@rkooo567 Are you comfortable with this PR on the whole ?

**Review by rkooo567 — APPROVED** [review]

**Review by sfc-gh-hazhang — APPROVED** [review]

**Review by sfc-gh-hazhang — APPROVED** [review]
